File: HW2/BasicTCPClient.c
/*
 * 20134220 Jeong hyun, Woo
 * BasicTCPClient.c
*/


#include <sys/socket.h>
#include <sys/stat.h>
#include <arpa/inet.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <signal.h>
#include <time.h>


#define MAXLINE 2048
#define IP "165.194.35.202" //nsl2.cau.ac.kr
#define PORT    24220

void printMenu();
void sigHandler(int sigNo);

int main(int argc, char** argv) {

    clock_t start,end,elapsed;
    struct sockaddr_in serveraddr;
    int option;
    int server_sockfd;
    int client_len;
    char buf[MAXLINE];
    char* tok;


    if((server_sockfd = socket(AF_INET, SOCK_STREAM, 0)) == -1) {
        perror("error");
        return 1;
    }

    serveraddr.sin_family = AF_INET;
    serveraddr.sin_addr.s_addr = inet_addr(IP);
    serveraddr.sin_port = htons(PORT);
    

    client_len = sizeof(serveraddr);

    if((connect(server_sockfd, (struct sockaddr*)&serveraddr, client_len)) == -1) {
        perror("connect error");
        return 1;
    }

    /* when user terminate program by Ctrl + c, print 'Bye bye~~' at terminal */
    signal(SIGINT,(void*)sigHandler);
    

    for(;;){
        /* option select */
        printMenu();
        memset(buf,0x00,MAXLINE);
        read(0,buf,MAXLINE);
        option = atoi(buf);
        
        /* If user input command is not in 1 ~ 5, Ask new input command */
        if(option < 1 || option > 5){
            printf("Invalid option : 1 ~ 5\n");
            continue;
        }

        /* If user input command is in 1 ~ 5 */
        switch(option) {
            /* command 1 : echo with lower -> upper */
            case 1 :
                start = clock(); 
                /* send command to server, in this case, 1 is send to server */
                if(write(server_sockfd, buf, 1) <= 0) {
                    perror("write error");
                    return 1;
                }
                end = clock(); 
                elapsed = end - start; // round-trip response time that spent sending command to the server

                /* get sentence */
                printf("Input lowercase sentence: ");
                fflush(stdout);
                memset(buf,0x00,MAXLINE);
                read(0,buf,MAXLINE);
                
                /* send to server */
                start = clock();
                if(write(server_sockfd, buf, strlen(buf)) <= 0) {
                    perror("write error");
                    return 1;
                }

                /* receive from server */
                memset(buf, 0x00, MAXLINE);
                if(read(server_sockfd, buf, MAXLINE) <= 0) {
                    perror("read error");
                    return 1;
                }

                end = clock();
                elapsed += end - start; // round-trip response time that spent sending and receiving sentence to the server
                
                /* print result */
                fprintf(stdout,"\nReply from server: %s",buf);
                fprintf(stdout,"Response time : %ld ms\n\n",elapsed);
                break;

            /* command 2 : get client's IP address & Port number */
            case 2 :
                start = clock();

                /* send command to the server, in this case, 2 is sent to server */
                if(write(server_sockfd, buf, 1) <= 0) {
                    perror("write error");
                    return 1;
                }
                
                /* recevie client's ip & port from server */
                memset(buf,0x00, MAXLINE);
                if(read(server_sockfd,buf,MAXLINE) <= 0) {
                    perror("ip,port read error");
                    return 1;
                }
                end = clock();
                elapsed = end - start; // round-trip response time
                
                /* print result 
                 * receive format : xxx.xxx.xxx.xxx,PORT -> split string with delimeter ','*/
                tok = strtok(buf,",");
                fprintf(stdout,"\nReply from server: IP=%s",tok);
                fflush(stdout);
                tok = strtok(NULL,",");
                fprintf(stdout," PORT=%s\n",tok);
                fprintf(stdout,"Response time : %ld ms\n\n",elapsed);
                break;

            /* command 3 : get server's current time */
            case 3 :
                start = clock();
                /* send command to the server, in this case, 3 is sent to server */
                if(write(server_sockfd, buf, 1) <= 0) {
                    perror("write error");
                    return 1;
                }
                /* receive the server's current time */
                memset(buf,0x00, MAXLINE);
                if(read(server_sockfd,buf,MAXLINE) <= 0) {
                    perror("read error");
                    return 1;
                }
                end = clock();
                elapsed = end - start; // round-trip response time

                /* print result */
                fprintf(stdout,"\nReply from server: %s\n",buf);
                fprintf(stdout,"Response time : %ld ms\n\n",elapsed);
                break;
            
            /* command 4 : get the server's service count */
            case 4 :
                start = clock();
                /* send command to the server, in this case, 4 is sent to server */
                if(write(server_sockfd, buf, 1) <= 0) {
                    perror("write error");
                    return 1;
                }

                /* receive the server's service count */
                memset(buf,0x00,MAXLINE);
                if(read(server_sockfd,buf,MAXLINE) <= 0) {
                    perror("read error");
                    return 1;
                }
                end = clock();
                elapsed = end - start; // round-trip response time

                /* print result */
                fprintf(stdout,"\nReply from server: %s\n",buf);
                fprintf(stdout,"Response time : %ld ms\n\n",elapsed);
                break;

            default :
                break;
        }

        
        /* if command is 5 or Ctrl + c, send command to server and terminate program */
        if(option == 5 || option == 0) {
            if(write(server_sockfd, buf, 1) <= 0) {
                    perror("write error");
                    return 1;
                }
            close(server_sockfd);
            printf("Bye bye~~\n");
            break;
        }
    }

    return 0;
}
void sigHandler(int sigNo) {
    printf("\nBye bye~~\n");
    exit(1);
}
void printMenu() {
    printf("<Menu>\n");
    printf("1) convert text to UPPER-case\n");
    printf("2) get my IP address and port number\n");
    printf("3) get server time\n");
    printf("4) get server service count\n");
    printf("5) exit\n");
    printf("Input option: ");
    fflush(stdout);
}

File: HW2/BasicTCPClient.py
#
# 20134220 Jeong hyun, Woo
# BasicTCPClient.py
#

from socket import *
import time
import sys

def printMenu() :
    print('<Menu>')
    print('1) convert text to UPPER-case')
    print('2) get my IP address and port number')
    print('3) get server time')
    print('4) get server service count')
    print('5) exit')

currentTime = lambda : int(time.time() * 1000)


serverName = '127.0.0.1'
serverPort = 24220

# create socket
clientSocket = socket(AF_INET, SOCK_STREAM)

# connect
clientSocket.connect((serverName, serverPort))
print("The client is running on port", clientSocket.getsockname()[1])

try :

    while True:

        printMenu()

        # User enter 'Ctrl + C' or 'Ctrl + Z' -> Program shutdown with 'Bye bye~~' message.
        try :
            option = input('Input option: ')
        except KeyboardInterrupt as e :
            clientSocket.close()
            print('Bye bye~~')
            break
        except EOFError as e :
            clientSocket.close()
            print('Bye bye~~')
            break

        try :
            # Option 1 : Convert text to UPPER-case.
            if int(option)== 1:
                
                # send command 
                start = currentTime()                                       #
                clientSocket.send(option.encode())                          # response time 1
                elapsed = currentTime() - start                             #

                message = input('Input lowercase sentece: ')
                
                # send sentence & receive uppercase-sentence
                start = currentTime()                                       #
                clientSocket.send(message.encode())                         # response time 2
                modifiedMessage = clientSocket.recv(2048)                   #
                
                elapsed += (currentTime() - start)                          # total response-time = response time1 + response time 2

                print('\nReply from server: ', modifiedMessage.decode())
                print('Response time:',elapsed,'ms\n')
            
            # Option2 : Get User IP address and port number.
            elif int(option) == 2:

                start = currentTime()
                # send command
                clientSocket.send(option.encode())
                # receive client ip & port number
                clientInfo = clientSocket.recv(2048)
            
                elapsed = currentTime() - start # round-trip time                   
                [ip, port] = clientInfo.decode().split(',')
                print('\nReply from server:','IP=',ip,'port=',port)
                print('Response time:',elapsed,'ms\n')
            
            # Option3 : Get server's current time.
            elif int(option) == 3:

                start = currentTime()
                # send command
                clientSocket.send(option.encode())
                # get server's current time
                serverTime = clientSocket.recv(2048)

                elapsed = currentTime() - start # round-trip time

                print('\nReply from server:', serverTime.decode())
                print('Response time:',elapsed,'ms\n')

            # Option4 : Get server's service count.
            elif int(option) == 4:

                start = currentTime()
                # send command
                clientSocket.send(option.encode())
                # receive server's service count
                serviceCount = clientSocket.recv(2048)

                elapsed = currentTime() - start # round-trip time

                print('\nReply from server:client request count =',serviceCount.decode())
                print('Response time:',elapsed,'ms\n')
            
            # Option5 : Program terminate.
            elif int(option) == 5:
                
                clientSocket.send(option.encode())
                print('Bye bye~')
                clientSocket.close()
                sys.exit(1)
            
            else :
                continue
        # if command is invalid form, get new input from user
        except ValueError :
            continue


except ConnectionAbortedError as e:
    clientSocket.close()
    print(e)
    sys.exit(1)

# if user terminate program by Ctrl + C
except KeyboardInterrupt :
    clientSocket.close()
    print('Bye bye~~')
    sys.exit(1)
